### PR TITLE
Action View Test Case `rendered` memoization

### PR DESCRIPTION
**Pull request by seanpdoyle**
### Motivation / Background

The introduction of memoization as an optimization posed a backwards incompatible change to View tests that call `render` multiple times.

### Detail

Follow-up to [49856][]
Follow-up to [49194][]

This commit changes the `@rendered` instance variable from a `String` to an instance of the `RenderedViewContent` specialized `String` subclass.
    
The end result is that there is no memoization to reset, and the memoization optimization side-effect is preserved after rendering for test cases where `rendered` (or parser methods like `rendered.html`) might be invoked more than once.

[49856]: https://github.com/rails/rails/pull/49856#issuecomment-1945039015
[49194]: https://github.com/rails/rails/pull/49194/files#diff-ce84a807f3491121a5230d37bd40454bb1407fcca71179e1a2fa76d4c0ddfa2aR293

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: actionview/lib/action_view/test_case.rb [hunk 1]
```diff
@@ -234,6 +234,7 @@ def config
       end
 
       def render(options = {}, local_assigns = {}, &block)
+        @_rendered_view_content = nil
```
Comment (seanpdoyle) [inline actionview/lib/action_view/test_case.rb:237]: @rafaelfranca Is this internal state reseting worth the maintenance cost when compared to removing the memoization instance variable entirely?

```diff
diff --git a/actionview/lib/action_view/test_case.rb b/actionview/lib/action_view/test_case.rb
index 97d1660200..006c6bf759 100644
--- a/actionview/lib/action_view/test_case.rb
+++ b/actionview/lib/action_view/test_case.rb
@@ -234,7 +234,6 @@ def config
       end
 
       def render(options = {}, local_assigns = {}, &block)
-        @_rendered_view_content = nil
         view.assign(view_assigns)
         @rendered << output = view.render(options, local_assigns, &block)
         output
@@ -292,7 +291,7 @@ def rendered_views
       #     assert_pattern { rendered.json => { title: "Hello, world" } }
       #   end
       def rendered
-        @_rendered_view_content ||= self.class.content_class.new(@rendered)
+        self.class.content_class.new(@rendered)
       end
 
       def _routes
@@ -389,7 +388,6 @@ def view
         :@_partials,
         :@passed,
         :@rendered,
-        :@_rendered_view_content,
         :@request,
         :@routes,
         :@tagged_logger,
```

File: actionview/lib/action_view/test_case.rb [hunk 2]
```diff
@@ -223,7 +223,7 @@ def setup_with_controller
         @request = @controller.request
         @view_flow = ActionView::OutputFlow.new
         @output_buffer = ActionView::OutputBuffer.new
-        @rendered = +""
+        @rendered = self.class.content_class.new(+"")
```
Comment (eugeneius) [inline actionview/lib/action_view/test_case.rb:226]: This breaks a compatibility workaround that was documented in the upgrading guide: 

https://github.com/rails/rails/blob/dd6f3e13285cff5e248340f30bcdc713eebed119/guides/source/upgrading_ruby_on_rails.md#L298-L308

Maybe we should go with clearing the ivar, since users may have added that workaround to their apps already.

Comment (seanpdoyle) [reply actionview/lib/action_view/test_case.rb:226]: Hmm, that workaround was suggested because the change to `rendered` returned an instance of a class that was not a `String`. 

The inheritance chain has changed to where the `content_class` is a `String` subclass, which might make that guidance no longer necessary.

**Comment by eugeneius** [comment]
Thank you Sean! Backported to 7-1-stable in https://github.com/rails/rails/commit/ddb959b2966140853827960da25fd1e2452ef2b6.

File: guides/source/upgrading_ruby_on_rails.md [hunk 3]
```diff
@@ -283,30 +283,6 @@ const fileInputSelector = Rails.fileInputSelector
 Rails.fileInputSelector(...)
 ```
 
-### `ActionView::TestCase#rendered` no longer returns a `String`
```
Comment (rafaelfranca) [inline guides/source/upgrading_ruby_on_rails.md:286]: Why was this removed?
